File: ITB2/src/DiffusionFilter.java
//funktioniert.

import itb2.filter.AbstractFilter;
import itb2.image.GrayscaleImage;
import itb2.image.Image;
import itb2.image.ImageFactory;

public class DiffusionFilter extends AbstractFilter{
	private static String EPS ="epsilon_0";
	private static String ITER = "number of iterations";
	private static String LAMBDA = "lambda";
	
	public DiffusionFilter() {
		this.properties.addDoubleProperty("epsilon_0", 1);
		this.properties.addIntegerProperty("number of iterations", 500);
		this.properties.addDoubleProperty("lambda", 1);
	}
	
	public Image filter(Image input) {
		double eps = properties.getDoubleProperty(EPS);
		int iter = properties.getIntegerProperty(ITER);
		double lambda = properties.getDoubleProperty(LAMBDA);
		double[][] image_t = Utility.imageToDoubleArray(input);
		double[][] image_t1 = new double[input.getHeight()][input.getWidth()];
		GrayscaleImage output = ImageFactory.bytePrecision().gray(input.getSize());
		
		for(int it =0; it<iter; it++) {
			double [][] dx = new double[input.getHeight()][input.getWidth()];
			double [][] dy = new double[input.getHeight()][input.getWidth()];
			double [][] flussx = new double[input.getHeight()][input.getWidth()];
			double [][] flussy = new double[input.getHeight()][input.getWidth()];
			double [][] dflussx = new double[input.getHeight()][input.getWidth()];
			double [][] dflussy = new double[input.getHeight()][input.getWidth()];
			double [][] divergenz = new double[input.getHeight()][input.getWidth()];
			
			
			//iteriere über Pixel, Randwerte ignoriert
			for (int row=1; row<input.getHeight()-1; row++) {
				for (int col=1; col<input.getWidth()-1; col++) {
					//Gradienten approximieren
					dx[row][col] = image_t[row+1][col]-image_t[row-1][col];
					dy[row][col] = image_t[row][col+1]-image_t[row][col-1];
					
					// Diffusionstensor berechnen
					double [][] diffTensor = {{eps*lambda*lambda/(eps*eps*(dx[row][col]*dx[row][col]+dy[row][col]*dy[row][col])+lambda*lambda), 0},
											  {0, eps*lambda*lambda/(eps*eps*dx[row][col]*dx[row][col]+eps*eps*dy[row][col]*dy[row][col]+lambda*lambda)}};
					// Berechen Fluss
					flussx[row][col] = -diffTensor[0][0]*dx[row][col];
					flussy[row][col] = -diffTensor[1][1]*dy[row][col];
					if(col==input.getWidth()-2) {
						col=col+1;
						col=col-1;
					}
				}
			}
			for (int row=1; row<input.getHeight()-1; row++) {
				for (int col=1; col<input.getWidth()-1; col++) {
					// Gradienten des Fluss schätzen
					dflussx[row][col]=flussx[row+1][col]-flussx[row-1][col];
					dflussy[row][col]=flussy[row][col+1]-flussy[row][col-1];
					
					// Berechne Divergenz pro Pixel
					divergenz[row][col]=dflussx[row][col]+dflussy[row][col];
					
					// berechen Ergebnisintensität
					image_t1[row][col] =image_t[row][col]-divergenz[row][col];
				}
			}
			// damit in nächster Iteration auf Werte der vorherigen Zugegriffen werden kann
			image_t=image_t1;
			
		}

		return Utility.doubleArrayToImage(image_t1, output, input.getWidth(), input.getHeight());
	}
}

File: ITB2/src/DiffusionFilter.kt
import itb2.filter.AbstractFilter
import itb2.image.GrayscaleImage
import itb2.image.Image
import itb2.image.ImageFactory
import kotlin.math.floor
import kotlin.math.pow
import kotlin.math.sqrt

 class DiffusionFilter : AbstractFilter() {

    /**
     * Returns double array of convolution results
     * Turns RGB Images into gray scale
     */
    fun aproximateDerivative(input: Array<DoubleArray>, width: Int, height: Int): Array<Array<DoubleArray>> {
        val result = Array(2) { Array(width) { DoubleArray(height) } }
        for (col in 0 until width) {
            for (row in 0 until height) {
                //default value for outside of the image
                val value = doubleArrayOf(0.0, 0.0)

                // only if indices are not beyond the edge calculation is valid
                if (!(col - 1 < 0 || col + 1 >= width)) {
                    value[0] = (input[col - 1][row] - input[col + 1][row])
                }
                if (!(row - 1 < 0 || row + 1 >= height)) {
                    value[1] = (input[col][row - 1] - input[col][row + 1])
                }
                result[0][col][row] = value[0]
                result[1][col][row] = value[1]
            }
        }
        return result
    }

    /**
     * Returns double array of convolution results
     * Turns RGB Images into gray scale
     */
    fun calculateDivergence(input: Array<Array<DoubleArray>>, width: Int, height: Int): Array<DoubleArray> {
        val result = Array(width) { DoubleArray(height) }
        for (col in 0 until width) {
            for (row in 0 until height) {
                //default value for outside of the image
                var value = 0.0

                // only if indices are not beyond the edge calculation is valid
                if (!(col - 1 < 0 || col + 1 >= width || row - 1 < 0 || row + 1 >= height)) {
                    value =
                        (input[col - 1][row] - input[col + 1][row])[0] + (input[col][row - 1] - input[col][row + 1])[1]
                }
                result[col][row] = value
            }
        }
        return result
    }


    /**
     * returns values of the absolute gradient
     */
    protected fun calculateFlow(
        horizontal: Array<DoubleArray>,
        vertical: Array<DoubleArray>,
        width: Int,
        height: Int,
        e0: Double,
        lambda: Double
    ): Array<Array<DoubleArray>> {
        val result = Array(width) { Array(height) { DoubleArray(2) } }

        //calulate absolute gradient and min max
        for (col in 0 until width) {
            for (row in 0 until height) {
                val tensorComponent = e0 * lambda.pow(2.0) / (lambda.pow(2.0) +
                    horizontal[col][row].pow(2.0) + vertical[col][row].pow(2.0)
                )
                result[col][row] =
                    doubleArrayOf(-tensorComponent * horizontal[col][row], -tensorComponent * vertical[col][row])
            }
        }
        return result
    }


    /**
     *
     * Converts image to gray scale
     */
    override fun filter(input: Image): Image {
        val grayImage = Grayfilter().filter(input)
        val output: Image = ImageFactory.bytePrecision().gray(input.size)
        val width = grayImage.width
        val height = grayImage.height
        val e0 = properties.getDoubleProperty(EPSILONZERO)
        val lambda = properties.getDoubleProperty(LAMBDA)
        val iterations = properties.getIntegerProperty(ITERATIONS)
        var values = Utility.imageToDoubleArray(input)
        for (i in 0 until iterations) {
            val result = Array(width) { DoubleArray(height) }
            val gradient = aproximateDerivative(values, width, height)
            val tensor = calculateFlow(gradient[0], gradient[1], width, height, e0, lambda)
            val divergence = calculateDivergence(tensor, width, height)
            for (col in 0 until width) {
                for (row in 0 until height) {
                    result[col][row] = values[col][row] - divergence[col][row]
                }
            }
            values = result
        }
        return Utility.doubleArrayToImage(values, output, width, height)
    }

    companion object {
        private const val EPSILONZERO = "epsilon zero"
        private const val LAMBDA = "lambda"
        private const val ITERATIONS = "iterations"
    }

    init {
        properties.addDoubleProperty(EPSILONZERO, 1.0)
        properties.addDoubleProperty(LAMBDA, 1.0)
        properties.addIntegerProperty(ITERATIONS, 1)
    }

    private operator fun DoubleArray.minus(substrate: DoubleArray): DoubleArray {
        assert(this.size == substrate.size)
        val result = DoubleArray(this.size)
        for ((i, v) in this.withIndex()) {
            result[i] = v - substrate[i]
        }
        return result
    }

    private operator fun DoubleArray.plus(substrate: DoubleArray): DoubleArray {
        assert(this.size == substrate.size)
        val result = DoubleArray(this.size)
        for ((i, v) in this.withIndex()) {
            result[i] = v + substrate[i]
        }
        return result
    }

}